### ISSUE TITLE
Fill shuffle buffer very slow

### ISSUE THREAD

**Issue by delock** [tensorflow/minigo#520]
When training with shuffle buffer size 100,000, it take a long time to fill us shuffle buffer, it can take up to 1 minute for each training epoch.

`
INFO:tensorflow:Saving checkpoints for 0 into /home/gma/results/minigo/final/estimator_working_dir/model.ckpt.                                                              
2018-10-15 00:01:20.219185: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:98] Filling up shuffle buffer (this may take a while): 23398 of 100000                     
2018-10-15 00:01:30.217862: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:98] Filling up shuffle buffer (this may take a while): 47305 of 100000                     
2018-10-15 00:01:40.218016: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:98] Filling up shuffle buffer (this may take a while): 70837 of 100000                     
2018-10-15 00:01:50.218091: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:98] Filling up shuffle buffer (this may take a while): 94768 of 100000                     
2018-10-15 00:01:52.428082: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:136] Shuffle buffer filled.                                                                
`

After some investigation, it turns out in preprocessing.py, a FilterDataset is used to sample input data:

`
dataset = dataset.filter(lambda x: tf.less(
    tf.random_uniform([1]), filter_amount)[0])  
`

When running this FilterDataset, it would run the predicate repeatedly, which is very slow.

I propose to introduce a 'SamplingDataset' to TensorFlow, which accepts a rate as parameter, and do sampling based on an internal random number generator.   It would greatly reduce the time spend on filling shuffle buffer.

**Comment by brilee** [comment]
Adding a new op to TensorFlow itself is outside the scope of the MiniGo project.

That being said - do you have profiler evidence to suggest that the inline filter is what is causing the slowness, rather than having to do the I/O for 20 examples for every 1 that ends up being used?

**Comment by delock** [comment]
I didn't successfully get profiling data with the following profiling code:
```
     with tf.contrib.tfprof.ProfileContext('/tmp/train_dir') as pctx:
         main.train(ESTIMATOR_WORKING_DIR, TRAINING_CHUNK_DIR, save_file,
                generation_num=model_num + 1)
```
But I verified by repeat calling EvaluatePredicate twice in tensorflow->filter_dataset_op.cc:
```
           TF_RETURN_IF_ERROR(
               dataset()->EvaluatePredicate(ctx, *out_tensors, &matched));
           TF_RETURN_IF_ERROR(
               dataset()->EvaluatePredicate(ctx, *out_tensors, &matched));
           if (!matched) {
             // Clear the output tensor list since it didn't match.
             out_tensors->clear();
```
The effect is the filter will be called twice every iteration.  It is purely filter overhead.  The amount of I/O will not change.   So its a quick verify whether the filter is the bottleneck.  And it turns out the filling up slowed down.  Note how it only fills up to 17118 for the first 10 second, while in my previous log, it fills up to 23398. 
```
2018-10-21 00:36:32.288609: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 17118 of 100000
2018-10-21 00:36:42.288490: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 34088 of 100000
2018-10-21 00:36:52.288549: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 50881 of 100000
2018-10-21 00:37:02.288416: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 67932 of 100000
2018-10-21 00:37:12.288681: I tensorflow/core/kernels/data/shuffle_dataset_op.cc:97] Filling up shuffle buffer (this may take a while): 84838 of 100000
```
I also experimented a sampling dataset in my local tensorflow branch.    This dataset takes a 'rate' parameter and do sampling inside the dataset op, instead of use filter to do the sampling.  The training is a lot faster:
Batch size = 128, filter amount = 0.05, board size 9x9, examples trained per generation: 100000
Measure training time:
Use filter dataset: 148.877 seconds
Use sampling dataset instead of filter dataset: 86.363 seconds
Use filter dataset (twice to double filter overhead): 197.128 seconds

**Comment by amj** [comment]
@delock we ended up migrating to using the Cloud Bigtable service, and the TPU can sample the data out of the bigtable directly; feel free to re-open this if you'd like to see something else done here.

**Comment by yerrick** [comment]
during my development, I found that "dataset.filter" for each element is slow. It's like the "dataset.map" that each time you run a python function in tensorflow, there will be static cost.

So the solution is to reduce the call of python function by "dataset.batch" before the "dataset.fillter". And the use "dataset.map" after "dataset.batch". 

in "dataset.map" you can use "tf.boolean_mask" to gather the element satisfy the filter condition.

my script is like this:

```
                mydataset = mydataset.batch(1000)
                mydataset = mydataset.map(_parse_function, num_parallel_calls=4)
                mydataset = mydataset.unbatch()

		def _parse_function(example_proto, filterfn=lambda x:xxxxxxxx):
			# Parse the input `tf.Example` proto using the dictionary above.
			parsedf = tf.io.parse_example(example_proto, feature_descriptiondebug)
			if filterfn is not None:
				filterbool = filterfn(parsedf)
				parsedf['fea'] = tf.boolean_mask(parsedf['fea'],filterbool)
				parsedf['label'] = tf.boolean_mask(parsedf['label'],filterbool)
			return parsedf
```

**Comment by tommadams** [comment]
Yeah, I experimented with a similar approach but ended up writing a C++ binary to perform shuffling and filtering: https://github.com/tensorflow/minigo/blob/master/cc/sample_records.cc

Selfplay generates a large number of small training records, and the `sample_records` binary batches them up into a small number large records, which also helps throughput when training. It reads and writes on multiple threads and so is pretty fast. We've only used it so far with the MLPerf benchmark so I don't know how well it would perform on a full run (like Andrew said, we use Cloud BigTable for those).

**Comment by delock** [comment]
A new C++ op is added to TensorFlow to support fast random sampling, and an optimization to convert the filter op automatically.
https://github.com/tensorflow/tensorflow/pull/25384

An example of applying the filter optimization can be found at
https://github.com/mlperf/training_results_v0.6/blob/master/Intel/benchmarks/minigo/implementations/tensorflow/preprocessing.py#L144 , in the case of small filter amount, performance improvement can easily be seen.

**Comment by tommadams** [comment]
It looks like that optimization has been disabled: https://github.com/tensorflow/tensorflow/blob/01981de67a0d509f1fb475d158f3ee852bc607be/tensorflow/core/grappler/optimizers/data/filter_with_random_uniform_fusion.cc#L275
